### DISCUSSION TITLE
Blazor - HttpClient Depencency Injection for HttpProxy

### DISCUSSION THREAD

**Discussion by rfcdejong** [MarimerLLC/csla#1979] [Questions]
I need to add a header on each request from Blazor to the server, but currently it feels wrong to me.

Current solution in Program.cs Main:

`builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });`

And the HttpClient is injected into the Csla HttpProxy. I think the current solution is to implement a custom HttpProxy and override GetHttpClient and do stuff there.

However I think it would be better [to allow HttpClientFactory](https://docs.microsoft.com/en-us/aspnet/core/blazor/call-web-api?view=aspnetcore-3.1#named-httpclient-with-ihttpclientfactory) or Typed HttpClient and use AddHttpClient and explicitly define it for a new class DataportalHttpClient which should be used in HttpProxy.

```
builder.Services.AddHttpClient<DataportalHttpClient>(client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
      .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();
```

This gives more flexibility, in this case I add BaseAddressAuthorizationMessageHandler, this ensures that an accesstoken is passed along to the server which will be validated beside trusting the deserialized claimsprincipal which could be manually modified.

**Comment by rfcdejong** [comment]
I did implement a tiny bit nicer solution, which works and passes along the jwt token in the authorization header.

```
    public class DataportalHttpProxy : HttpProxy
    {
        private readonly HttpClient _httpClient;

        public DataportalHttpProxy(IHttpClientFactory clientFactory)
        {
            _httpClient = clientFactory.CreateClient("ServerAPI");
        }

        protected override HttpClient GetHttpClient()
        {
            return _httpClient;
        }
    }
```

and in Program.cs

```
     //            builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });

     builder.Services.AddHttpClient();
     builder.Services.AddHttpClient("ServerAPI", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
         .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();

     builder.UseCsla(c => c.DataPortal().DefaultProxy(typeof(DataportalHttpProxy), "/api/DataPortal"));
```

**Reply by rockfordlhotka** [reply]
I can't argue with the use of `IHttpClientFactory`, other than that the Blazor project templates all set up `HttpClient` for use.

I want to support the "easy F5 path", which means supporting `HttpClient`. I think that's important.

At the same time, I think you are right. It is important to support the superior `IHttpClientFactory` model.

Maybe the answer is to do what you've done, and include two "HttpProxy" types, one for each scenario?

**Reply by mtavares628** [reply]
Is there any other code that's missing to make that work?  I've created the same DataportalHttpProxy class, and I was under the impression that the AddHttpClient() calls would be enough to inject the IHttpClientFactory into the constructor of the derived proxy class, but when I run it I'm getting an error saying: Cannot create instance of Type 'DataportalHttpProxy'. No public parameterless constructor found.  Do I need to register the DataportalHttpProxy class as well?  Here's what my Program.cs looks like:
```

public static async Task Main(string[] args)
        {
            var builder = WebAssemblyHostBuilder.CreateDefault(args);
            builder.RootComponents.Add<App>("#app");

            //builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });

            builder.Services.AddHttpClient();
            builder.Services.AddHttpClient("DadApi", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();

            builder.Services.AddOidcAuthentication(options =>
            {
                builder.Configuration.Bind("oidc", options.ProviderOptions);
            });
            App.IsServerSide = false;

           builder.Services.AddAuthorizationCore();
           builder.Services.AddOptions();
           builder.Services.AddSingleton<CslaUserService>();

            builder.UseCsla(c => c
                .DataPortal()
                .DefaultProxy(typeof(DataportalHttpProxy),
                              "https://localhost:6001/api/dataportaltext/")
                );
            
            await builder.Build().RunAsync();
        }
```

Any insight would be greatly appreciated.
